firmware/common/fpga.c

```diff
@@ -22,226 +22,76 @@
+		if ((drv->regs_dirty >> r) & 0x1) {
+			fpga_reg_commit(drv, r);
 		}
-		seq = lfsr_advance(seq);
 	}
-
-	// Update selftest result.
-	selftest.sgpio_rx_ok = seq_in_sync;
-	if (!selftest.sgpio_rx_ok) {
-		selftest.report.pass = false;
-	}
-
-	return selftest.sgpio_rx_ok;
 }
 
-bool fpga_if_xcvr_selftest()
+void fpga_hw_sync_enable(fpga_driver_t* const drv, const hw_sync_mode_t hw_sync_mode)
 {
-#if defined(DFU_MODE) || defined(RAM_MODE)
-	return false;
-#endif
-
-	// Skip if FPGA configuration failed.
-	if (!selftest.fpga_image_load_ok) {
-		selftest.xcvr_loopback_ok = false;
-		return false;
-	}
-
-	const size_t num_samples = USB_BULK_BUFFER_SIZE / 2;
-
-	// Set gateware features for the test.
-	ssp1_set_mode_ice40();
-	ice40_spi_write(&ice40, 0x01, 0x1); // RX DC block
-	ice40_spi_write(&ice40, 0x05, 128); // NCO phase increment
-	ice40_spi_write(&ice40, 0x03, 1);   // NCO TX enable
-	ssp1_set_mode_max283x();
-
-	// Configure RX calibration path and settle for 1ms.
-	rf_path_set_direction(&rf_path, RF_PATH_DIRECTION_RX_CALIBRATION);
-	delay_us_at_mhz(1000, 204);
-
-	// Stream samples from the FPGA.
-	m0_set_mode(M0_MODE_RX);
-	m0_state.shortfall_limit = 0;
-	baseband_streaming_enable(&sgpio_config);
-	while (m0_state.m0_count < num_samples)
-		;
-	baseband_streaming_disable(&sgpio_config);
-	m0_set_mode(M0_MODE_IDLE);
-
-	rf_path_set_direction(&rf_path, RF_PATH_DIRECTION_OFF);
-
-	// Gateware default settings.
-	ssp1_set_mode_ice40();
-	ice40_spi_write(&ice40, 0x01, 0);
-	ice40_spi_write(&ice40, 0x03, 0);
-	ssp1_set_mode_max283x();
-
-	// Count zero crossings in the received samples.
-	// N/2 samples/channel * 2 zcs/cycle / 8 samples/cycle = N/8 zcs/channel
-	unsigned int expected_zcs = num_samples / 8;
+	fpga_reg_read(drv, FPGA_STANDARD_CTRL);
+	set_FPGA_STANDARD_CTRL_TRIGGER_EN(drv, hw_sync_mode == 1);
+	fpga_regs_commit(drv);
+}
 
-	unsigned int zcs_i = 0;
-	unsigned int zcs_q = 0;
-	uint8_t last_sign_i = 0;
-	uint8_t last_sign_q = 0;
-	for (size_t i = 0; i < num_samples; i += 2) {
-		uint8_t sign_i = (usb_bulk_buffer[i] & 0x80) ? 1 : 0;
-		uint8_t sign_q = (usb_bulk_buffer[i + 1] & 0x80) ? 1 : 0;
-		zcs_i += sign_i ^ last_sign_i;
-		zcs_q += sign_q ^ last_sign_q;
-		last_sign_i = sign_i;
-		last_sign_q = sign_q;
+void fpga_prbs_enable(fpga_driver_t* const drv, const bool enable)
+{
+	set_FPGA_STANDARD_CTRL(drv, 0);
+	if (enable) {
+		set_FPGA_STANDARD_CTRL_PRBS(drv, enable);
 	}
+	fpga_regs_commit(drv);
+}
 
-	// Allow a zero crossings counting error of +-5%.
-	bool i_in_range = (zcs_i > expected_zcs * 0.95) && (zcs_i < expected_zcs * 1.05);
-	bool q_in_range = (zcs_q > expected_zcs * 0.95) && (zcs_q < expected_zcs * 1.05);
-
-	// Update selftest result.
-	selftest.xcvr_loopback_ok = i_in_range && q_in_range;
-	if (!selftest.xcvr_loopback_ok) {
-		selftest.report.pass = false;
+void fpga_nco_enable(fpga_driver_t* const drv, const bool enable)
+{
+	if (enable) {
+		set_FPGA_STANDARD_CTRL_DC_BLOCK(drv, 1);  // RX DC block
```

Comment:
Enabling the TX NCO does not involve enabling RX DC block

Reply:
Well, I wouldn't join both features in a single one. In the selftest, we enable DC block and then enable TX NCO.